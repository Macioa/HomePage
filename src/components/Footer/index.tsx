import React from 'react'

const style = require('../../styles.css')

interface Link {
  name: string
  logo: any
  url: string
}

export const footerLinks = [
  {
    name: 'Linked In',
    logo: require('Images/linkedin.svg'),
    url: 'https://linkedin.com/in/ryanwademontgomery'
  },
  {
    name: 'General Assembly',
    logo: require('Images/ga.svg'),
    url: 'https://profiles.generalassemb.ly/ryan-montgomery'
  },
  {
    name: 'Twitter',
    logo: require('Images/twitter.svg'),
    url: 'https://twitter.com/macioa_macioa'
  },
  {
    name: 'Github',
    logo: require('Images/github.svg'),
    url: 'https://github.com/Macioa'
  },
  {
    name: 'Mail',
    logo: require('Images/mail.svg'),
    url: 'mailto:development@ryanwademontgomery.com'
  }
]
  .filter(e => e.logo && e.url)
  .map((e, i) => (
    <a
      id={`Link${e.name.replace(' ', '')}`}
      key={`Link${i}`}
      href={e.url}
      target='blank'>
      <img src={e.logo} className={style.footerLink} alt={e.name} />
    </a>
  ))

export const Footer = () => (
  <footer className={style.footer} id={style.footer}>
    {...footerLinks}
  </footer>
)
