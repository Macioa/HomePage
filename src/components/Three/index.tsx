import React, { useRef, useEffect } from 'react'
import {
  Scene,
  PerspectiveCamera,
  Renderer,
  Mesh,
  WebGLRenderer,
  Vector3,
  Geometry,
  TextGeometry,
  Material,
  MeshBasicMaterial,
  Font
} from 'three'
import { World, Body, Vec3, Sphere, Box } from 'cannon'

export interface Instance {
  world?: World
  gravity?: Vec3
  scene?: Scene
  camera?: PerspectiveCamera
  renderer?: Renderer
  objects?: Obj[]
}

export interface Obj {
  mesh: Mesh
  body?: Body
  shape?: string
  startpos: number[]
  mass: number
  anchorgrav?: number
}

// create necessary objects for scene
export const init = (
  {
    world = new World(),
    gravity = new Vec3(0, 0, -9.82),
    scene = new Scene(),
    camera = new PerspectiveCamera(
      75,
      window.innerWidth / window.innerHeight,
      0.1,
      1000
    ),
    renderer = new WebGLRenderer()
  }: Instance = {},
  camerapos = [0, 0, 45]
): Instance => {
  camera.position.set(camerapos[0], camerapos[1], camerapos[2])
  world.gravity = gravity
  return { world, gravity, scene, camera, renderer }
}

// resize render and reproject aspect ratio for camera
export const Resize = (i: Instance, w: number, h: number): void => {
  i.renderer.setSize(w, h)
  Reproject(i, w, h)
}

// reproject asspect ratio for camera
export const Reproject = ({ camera }: Instance, w: number, h: number): void => {
  camera.aspect = w / h
  camera.updateProjectionMatrix()
  camera.position.setZ((175 * h) / w)
}

// start render
export const Start = (i: Instance): void => {
  var prevRender: number
  const animate = (t = Date.now(), step = 1 / 60, maxSubSteps = 3) => {
    requestAnimationFrame(animate)
    i.renderer.render(i.scene, i.camera)
    if (prevRender !== undefined) {
      var dt = (t - prevRender) / 1000
      i.world.step(step, dt, maxSubSteps)
    }
    prevRender = t
    if (i.objects)
      i.objects.forEach(o => {
        // @ts-ignore
        o.mesh.position.copy(o.body.position)
        // @ts-ignore
        o.mesh.setRotationFromQuaternion(o.body.quaternion)
      })
  }

  animate()
}

// create geometry from string
const mem = new Map()
export const ComputeTextGeometry = (text = '', font: Font): Geometry[] => {
  return text.split('').map(c => {
    let data = mem.get(c)
      ? mem.get(c)
      : new TextGeometry(c, {
          size: 10,
          height: 0.25,
          curveSegments: 3,
          font: font
        })
    if (!mem.get(c)) data.computeBoundingBox()
    mem.set(c, data)
    return data
  })
}

// create meshes from geometry
export const CreateMeshes = (
  geos: Geometry[] = [],
  material: Material
): Mesh[] => geos.map(g => new Mesh(g, material))

// add elements to simulation //compute bounds, find largest bound, create CANNON body, add to THREE scene and CANNON sim
export const AddToSim = (
  i: Instance,
  meshes: Obj[] = [],
  defaultMass: number = 1000
): Obj[] => {
  if (!i.objects) i.objects = []
  return meshes.map(({ mesh, shape, startpos, mass, anchorgrav = null }) => {
    let size = mesh.geometry.boundingBox.getSize(new Vector3()),
      maxBound = Object.values(size).sort((a, b) => b - a)[0],
      sp = startpos.length >= 3 ? startpos : [0, 0, 0]
    mass = mass ? mass : defaultMass
    shape = shape.toLowerCase() === 'sphere' ? 'sphere' : 'box'
    let cShape =
        shape === 'sphere'
          ? new Sphere(maxBound)
          : new Box(new Vec3(size.x, size.y, size.z)),
      body = new Body({
        mass,
        position: new Vec3(sp[0], sp[1], sp[2]),
        shape: cShape
      })
    i.world.addBody(body)
    i.scene.add(mesh)
    let res = { mesh, body, shape, startpos, mass, anchorgrav }
    i.objects.push(res)
    if (res.anchorgrav) {
      res.body.preStep = () => {
        //  vel + (startingpos-currentpos) * anchorgrav
        res.body.velocity = res.body.velocity.vadd(
          new Vec3(res.startpos[0], res.startpos[1], res.startpos[2])
            .vadd(res.body.position.negate())
            .scale(res.anchorgrav)
        )
        // torq - eulerAngle
        let q = new Vec3()
        res.body.quaternion.toEuler(q)
        res.body.torque = res.body.torque.vsub(q)
      }
    }
    return res
  })
}

// create full objects from string
export const TextMesh = ({
  text = 'Hello World',
  spacing = 8,
  font = new Font(require('./fontdata/Moonglade.json')),
  material = new MeshBasicMaterial({ color: 0xffffff }),
  options = {}
}): Obj[] => {
  let meshes = CreateMeshes(ComputeTextGeometry(text, font), material).map(
    m => {
      return {
        mesh: m,
        width: m.geometry.boundingBox.getSize(new Vector3()).x
      }
    }
  )
  const sum = (total: number, mesh: any) => total + mesh.width + spacing
  let totalWidth = meshes.reduce(sum, spacing),
    currentpos = -totalWidth / 2 // left most bound
  return meshes.map(m => {
    let res = {
      mesh: m.mesh,
      shape: 'box',
      mass: 1,
      startpos: [currentpos + m.width / 2, 0, 0],
      anchorgrav: 0.2,
      ...options
    }
    currentpos = currentpos + m.width + spacing
    return res
  })
}

//
//
//
// create react element
export const ThreeCanvas = (props: any) => {
  const ref = useRef(null)
  let i = init({ gravity: new Vec3(0, 0, 0) })
  const resolve = () =>
    Resize(i, ref.current.offsetWidth, ref.current.offsetHeight)
  useEffect(() => {
    resolve()
    ref.current.appendChild(i.renderer.domElement)
    window.addEventListener('resize', resolve)

    AddToSim(i, TextMesh({ text: 'Ryan Montgomery' }))

    Start(i)
  })
  return (
    <div
      ref={ref}
      style={{ width: '100%', height: '100%' }}
      id='3Scene'
      {...props}
    />
  )
}
