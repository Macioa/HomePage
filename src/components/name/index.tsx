import React, { useRef, useMemo } from 'react'
import ReactDOM from 'react-dom'
import * as three from 'three'
import * as cannon from 'cannon'
import './styles.css'
import * as style from './styles.css'

export const ThreeCanvas = () => {
  const t = three,
    c = cannon

  return <div>test</div>
}
