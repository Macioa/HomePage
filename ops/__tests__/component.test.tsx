import React from 'react'
import renderer from 'react-test-renderer'
import { Hello } from '../../src/components/hello'

describe('Addition', () => {
  it('knows that 2 and 2 make 4', () => {
    expect(2 + 2).toBe(4)
  })
})

describe('Todo component renders the todo correctly', () => {
  it('renders correctly', () => {
    const rendered = renderer.create(
      <Hello compiler='test/ts' framework='test/react' />
    )
    expect(rendered.toJSON()).toMatchSnapshot()
  })
})

// describe('ThreeJS Canvas component renders', () => {
//   it('renders correctly', () => {
//     const rendered = renderer.create(<ThreeCanvas />)
//     expect(rendered.toJSON()).toMatchSnapshot()
//   })
// })
