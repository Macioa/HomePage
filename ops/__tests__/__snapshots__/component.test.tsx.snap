// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer Component Matches snapshot 1`] = `
<footer>
  <a
    href="https://linkedin.com/in/ryanwademontgomery"
    id="LinkLinkedIn"
    target="blank"
  >
    <img
      alt="Linked In"
      src="div"
    />
  </a>
  <a
    href="https://profiles.generalassemb.ly/ryan-montgomery"
    id="LinkGeneralAssembly"
    target="blank"
  >
    <img
      alt="General Assembly"
      src="div"
    />
  </a>
  <a
    href="https://twitter.com/macioa_macioa"
    id="LinkTwitter"
    target="blank"
  >
    <img
      alt="Twitter"
      src="div"
    />
  </a>
  <a
    href="https://github.com/Macioa"
    id="LinkGithub"
    target="blank"
  >
    <img
      alt="Github"
      src="div"
    />
  </a>
  <a
    href="mailto:development@ryanwademontgomery.com"
    id="LinkMail"
    target="blank"
  >
    <img
      alt="Mail"
      src="div"
    />
  </a>
</footer>
`;
